rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,170 @@
+the file.
+
+Is it safe to assume/require the file to be encrypted using AES, or should we
+support other algorithms as well?
+
+*criteria*
+
+`criteria` is a single level, key/value object that specifies the fields and
+values to lookup. The fields must all match (AND) for the row to be returned.
+
+*case_insensitive*
+
+By default the search will be case sensitive, but this can be changed by 
+passing `true` to this parameter.
+
+## Internal Proposal
+
+- Describe your change as if you were presenting it to the Vector team.
+- Use lists, examples, and code blocks for efficient reading.
+- Be specific!
+
+The majority of the changes will be provided by adding two new functions within
+`vrl/stdlib` and some shared lookup code.
+
+Since the CSV file is loaded at bootime, it is not essential that it is loaded 
+using async. This makes things much simpler as currently the VRL compilation 
+process is entirely synchronous. No code outside of the new function in the
+vrl stdlib should need changing.
+
+The entire data file will be loaded into memory, so all lookups will be 
+performed in memory.
+
+### Indexing 
+
+Although the first version is not going to be doing any indexing into the data
+it is worth bearing it in mind as we will most likely need to add this in due
+course.
+
+In order to perform the indexing VRL needs to know which fields to index. The 
+criteria is being passed in as an object. If the type def for that object is
+known at runtime, we can extract the fields from this. 
+
+If the type def isn't known we could 
+
+- raise an error 
+- perform the search unindexed
+- create the index at runtime - generally the shape of this object will not 
+  change over time, so most likely the cost of creating the index would only
+  need to be paid on the first event.
+
+The approach we take here should be decided before the first release so we don't
+have to introduce a breaking change further down the line.
+
+Actual indexing strategies can be decided later.
+
+## Rationale
+
+There is significant customer demand for this feature.
+
+## Drawbacks
+
+1. There are a number of issues with parsing CSV files in particular around
+   handling delimeters and separators.
+2. Performance. Whilst we can ensure all the IO is performed at boot time,
+   searching through that data could still be quite expensive.
+3. Memory use. Since the data file is loaded into memory, the data will use up 
+   some memory. If the file is large enough that could have an impact should
+   Vector be running in a constrained environment.
+
+## Alternatives
+
+### Use a predicate for the search
+
+Instead of using an object to specify the search criteria we could allow the
+user to specify a predicate to determine the row to use for enrichment.
+
+```
+lookup_csv_row("/path/to/file.csv", |row| row.some_key = .some_field)
+```
+
+Note the ability to specify closures this is not yet available in VRL.
```

Comment:
I would not favour this right now, at least not until we've fleshed out #6031 and #3740.

Reply:
Yeah, agreed, it seems like this feature will ship first.